chainer/links/connection/sgu.py

```diff
@@ -0,0 +1,136 @@
+import numpy
+
+import chainer
+from chainer import cuda
+from chainer.functions.activation import hard_sigmoid
+from chainer.functions.activation import sigmoid
+from chainer.functions.activation import softplus
+from chainer.functions.activation import tanh
+from chainer.functions.math import linear_interpolate
+from chainer import link
+from chainer.links.connection import linear
+from chainer import variable
+
+
+class SGU(link.Chain):
+
+    def __init__(self, in_size, out_size):
+        super(SGU, self).__init__(
+            W_xh=linear.Linear(in_size, out_size),
+            W_zxh=linear.Linear(out_size, out_size),
+            W_xz=linear.Linear(in_size, out_size),
+            W_hz=linear.Linear(out_size, out_size),
+        )
+
+    def __call__(self, h, x):
+        x_g = self.W_xh(x)
+        z_g = tanh.tanh(self.W_zxh(x_g * h))
+        z_out = softplus.softplus(z_g * h)
+        z_t = hard_sigmoid.hard_sigmoid(self.W_xz(x) + self.W_hz(h))
+        h_t = linear_interpolate.linear_interpolate(z_t, z_out, h)
+        return h_t
+
+
+class StatefulSGU(SGU):
```

Comment:
I think stateful RNN should not be a derived class of stateless one because if `x` is an instance of `StatefulGRU`, `isinstance(x, SGU)` equals to `True`, which is counterintuitive.

Actually, existing RNN units, use base class like `LSTMBase` and `GRUBase`. Both stateful and stateless RNNs derive this base class.